### PR TITLE
akka.cluster.sharding - batch update

### PR DESCRIPTION
**Pull request by zbynek001**
Mostly Sharding allocation / rebalancing changes making use of the newly introduced AppVersion.
Let me know if you want to split it into smaller parts.

* Holistic shard allocation strategy,  (migrated from https://github.com/akka/akka/pull/29555)

  *  The rebalance in the LeastShardAllocationStrategy is only comparing the region
  with most shards with the one with least shards. Makes the rebalance rather
  slow. By default it's only rebalancing 1 shard at a time.
  * This new strategy looks at all current allocations to find the optimal
  number of shards per region and tries to adjust towards that value.
  Picking from all regions with more shards than the optimal.
  * Absolute and relative limit on how many shards that can be rebalanced
  in one round.
  * It's also not starting a new rebalance round until the previous has
  completed.
  * unit tests
  * second phase for fine grained rebalance, due to rounding it will not be perfect in the first phase
  * randomized unit test
  * configuration settings
  * docs

* Cluster aware shard allocation and rolling updates (migrated from https://github.com/akka/akka/pull/29548)
  * Adds some level of cluster awareness to both the LeastShardAllocationStrategy implementations:
  * prefer shard allocations on new nodes during rolling updates
  * don't rebalance during rolling update
  * don't rebalance when there are joining nodes
  * don't allocate to leaving, downed, exiting and unreachable nodes
  * When allocating when there are joining, unreachable, are leaving are de-prioritized to decrease the risk that a shard is allocated just to directly need to be re-allocated on a different node.

* Handle when there are no known regions yet (migrated from https://github.com/akka/akka/pull/29712)

* ShardRegion ignores HostShard when shutting down (migrated from https://github.com/akka/akka/pull/29707)

* Adds ActorRef.Ignore (Migrated from https://github.com/akka/akka/pull/28630)

* Allocate rebalanced shards immediately (migrated from https://github.com/akka/akka/pull/29551)
  * Otherwise it will continue to deallocate when there is no traffic
  that triggers allocations

* Forward terminated from ShardCoordinator to RebalanceWorker (migrated from https://github.com/akka/akka/pull/29463)
  * Avoiding the need for rebalance workers to watch shard regions which is
expensive as there is one rebalance worker per shard
  * fix race in rebalance worker (partially migrated from https://github.com/akka/akka/pull/29502)

* ClusterSharding: increasing retry interval for registration call (migrated from https://github.com/akka/akka/pull/26756)

* Make a distinction between hand-offs for rebalance and region shutdown (migrated from https://github.com/akka/akka/pull/29579)

* improve logging of shard region shutdown timeout (migrated from https://github.com/akka/akka/pull/29747)

### CODE REVIEW

File: src/contrib/cluster/Akka.Cluster.Sharding/ShardRegion.cs [hunk 1]
```diff
@@ -378,8 +440,11 @@ public ShardRegion(string typeName, Func<string, Props> entityProps, ClusterShar
             _replicator = replicator;
             _majorityMinCap = majorityMinCap;
 
-            _retryTask = Context.System.Scheduler.ScheduleTellRepeatedlyCancelable(Settings.TunningParameters.RetryInterval, Settings.TunningParameters.RetryInterval, Self, Retry.Instance, Self);
             SetupCoordinatedShutdown();
+
+            _retryInterval = Settings.TuningParameters.RetryInterval;
+            _initRegistrationDelay = TimeSpan.FromMilliseconds(100).Max(new TimeSpan(_retryInterval.Ticks / 2 / 2 / 2));
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding/ShardRegion.cs:446]: why `_retryInterval.Ticks / 2 / 2 / 2` ?

File: src/contrib/cluster/Akka.Cluster.Sharding/ShardRegion.cs [hunk 2]
```diff
@@ -718,25 +809,63 @@ private void HandleShardRegionCommand(IShardRegionCommand command)
...
 
+                    // we depend on the coordinator each time, if empty we need to register
+                    // otherwise we can try to deliver some buffered messages
                     if (_coordinator == null) Register();
                     else
                     {
-                        RequestShardBufferHomes();
+                        // Note: we do try to deliver buffered messages even in the middle of
+                        // a graceful shutdown every message that we manage to deliver is a win
+                        TryRequestShardBufferHomes();
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding/ShardRegion.cs:825]: LGTM

File: src/core/Akka/Actor/ActorPath.cs [hunk 3]
```diff
@@ -445,7 +445,7 @@ public string ToStringWithoutAddress()
         /// <inheritdoc/>
         public override string ToString()
         {
-            return ToStringWithAddress();
+            return $"{Address}{Join()}";
```
Comment (Aaronontheweb) [inline src/core/Akka/Actor/ActorPath.cs:448]: Might need to check performance impact here - making a note

File: src/core/Akka/Actor/ActorRef.cs [hunk 4]
```diff
@@ -569,6 +569,68 @@ public override bool IsLocal
         public override bool IsTerminated { get { return false; } }
     }
 
+
+    /// <summary>
+    /// An ActorRef that ignores any incoming messages.
```
Comment (Aaronontheweb) [inline src/core/Akka/Actor/ActorRef.cs:574]: Very cool

File: src/core/Akka/Actor/ActorRefProvider.cs [hunk 5]
```diff
@@ -53,6 +53,8 @@ public interface IActorRefProvider
         /// <summary>Gets the dead letters.</summary>
         IActorRef DeadLetters { get; }
 
+        IActorRef IgnoreRef { get; }
```
Comment (Aaronontheweb) [inline src/core/Akka/Actor/ActorRefProvider.cs:56]: LGTM

File: src/core/Akka.Tests/Actor/ActorRefIgnoreSpec.cs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+using Akka.Serialization;
+using Akka.TestKit;
+using Akka.TestKit.TestActors;
+using Xunit;
+using Akka.Util.Internal;
+using FluentAssertions;
+
+namespace Akka.Tests.Actor
+{
+    public class ActorRefIgnoreSpec : AkkaSpec, INoImplicitSender
```
Comment (Aaronontheweb) [inline src/core/Akka.Tests/Actor/ActorRefIgnoreSpec.cs:21]: LGTM

File: src/core/Akka.Cluster.Tests/ActorRefIgnoreSerializationSpec.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+        protected override void AfterAll()
+        {
+            base.AfterAll();
+            system1.Terminate();
+            system2.Terminate();
+        }
+
+        [Fact]
+        public void ActorSystem_IgnoreRef_should_return_a_serializable_ActorRef_that_can_be_sent_between_two_ActorSystems_using_remote()
```
Comment (Aaronontheweb) [inline src/core/Akka.Cluster.Tests/ActorRefIgnoreSerializationSpec.cs:54]: LGTM

**Review by Aaronontheweb — CHANGES_REQUESTED** [review]
Have a handful of minor changes and comments

File: src/contrib/cluster/Akka.Cluster.Sharding.Tests.MultiNode/MultiNodeClusterShardingConfig.cs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            AdditionalConfig = additionalConfig;
+            Loglevel = loglevel;
+
+            Config persistenceConfig = (mode == ClusterShardingSettings.StateStoreModeDData) ? ConfigurationFactory.Empty : PersistenceConfig();
+
+            Config common =
+                ConfigurationFactory
+                    .ParseString($@"""
+                        akka.actor.provider = ""cluster""
+                        akka.cluster.downing-provider-class = akka.cluster.testkit.AutoDowning
```
[the remark this answers is not in the document]
Comment (zbynek001) [reply src/contrib/cluster/Akka.Cluster.Sharding.Tests.MultiNode/MultiNodeClusterShardingConfig.cs]: removed, leftover from akka code

File: src/contrib/cluster/Akka.Cluster.Sharding/Internal/AbstractLeastShardAllocationStrategy.cs [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+{
+    using static Akka.Cluster.ClusterEvent;
+    using ShardId = String;
+
+    /// <summary>
+    /// Common logic for the least shard allocation strategy implementations
+    /// </summary>
+    public abstract class AbstractLeastShardAllocationStrategy : IActorSystemDependentAllocationStrategy
+    {
+        private static ImmutableHashSet<MemberStatus> JoiningCluster = ImmutableHashSet.Create(MemberStatus.Joining, MemberStatus.WeaklyUp);
```
[the remark this answers is not in the document]
Comment (zbynek001) [reply src/contrib/cluster/Akka.Cluster.Sharding/Internal/AbstractLeastShardAllocationStrategy.cs]: changed

File: src/contrib/cluster/Akka.Cluster.Sharding/Internal/AbstractLeastShardAllocationStrategy.cs [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+        private static ImmutableHashSet<MemberStatus> LeavingClusterStatuses = ImmutableHashSet.Create(MemberStatus.Leaving, MemberStatus.Exiting, MemberStatus.Down);
```
[the remark this answers is not in the document]
Comment (zbynek001) [reply src/contrib/cluster/Akka.Cluster.Sharding/Internal/AbstractLeastShardAllocationStrategy.cs]: changed

File: src/contrib/cluster/Akka.Cluster.Sharding/ShardRegion.cs (hunk 1 above)
Comment (zbynek001) [reply src/contrib/cluster/Akka.Cluster.Sharding/ShardRegion.cs:446]: this is to make the registration retry process faster at the beginning and then the retry interval is doubled until it matches the normal `_retryInterval` over 3 iterations

File: src/core/Akka/Actor/ActorPath.cs (hunk 3 above)
Comment (zbynek001) [reply src/core/Akka/Actor/ActorPath.cs:448]: should not, the previous `ToStringWithAddress` is doing the same inside

File: src/contrib/cluster/Akka.Cluster.Sharding.Tests.MultiNode/MultiNodeClusterShardingConfig.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Aaronontheweb) [reply src/contrib/cluster/Akka.Cluster.Sharding.Tests.MultiNode/MultiNodeClusterShardingConfig.cs]: Ah, thanks for clarifying

File: src/contrib/cluster/Akka.Cluster.Sharding/ShardRegion.cs (hunk 1 above)
Comment (Aaronontheweb) [reply src/contrib/cluster/Akka.Cluster.Sharding/ShardRegion.cs:446]: I see

**Pull request merged**